Adding support for ROCm for the LBANN software stack.

bvanessen opened the pull request:
depends on #21822

bvanessen committed:
Adding support for ROCm for the LBANN software stack.

haampie commented on var/spack/repos/builtin/packages/aluminum/package.py:
```diff
@@ -37,13 +37,18 @@ class Aluminum(CMakePackage, CudaPackage):
     depends_on('cub', when='@:0.1,0.6.0: +cuda ^cuda@:10.99')
+    depends_on('cub', when='@:0.1,0.6.0: +rocm')
+
+    conflicts('+cuda', when='+rocm')
+    conflicts('+rocm', when='+cuda')
```
```suggestion
    conflicts('+cuda', when='+rocm', msg='CUDA and ROCm support are mutually exclusive')
```
One `conflicts` is enough, and it can't hurt to add an error message :)

haampie commented on var/spack/repos/builtin/packages/dihydrogen/package.py:
```diff
@@ -77,10 +77,16 @@ class Dihydrogen(CMakePackage, CudaPackage):
 
     for arch in CudaPackage.cuda_arch_values:
         depends_on('aluminum cuda_arch=%s' % arch, when='+al +cuda cuda_arch=%s' % arch)
 
```
should this have a `conflicts('+cuda', when='+rocm', msg='CUDA and ROCm support are mutually exclusive')` too?

bvanessen replied:
I don't know how to set a conflicts between the `cuda_arch` and `amdgpu_target` variants since they are not just boolean flags.  I would assume that I want something like `conflicts('cuda_arch~=none', when='amdgpu_target~=none')` but I am not sure that will work.

haampie replied:
I think there's already conflicts in CudaPackage and ROCmPackage that handle this; you can't set a cuda target when ~cuda it set etc.

haampie commented on var/spack/repos/builtin/packages/hydrogen/package.py:
```diff
@@ -63,6 +63,8 @@ class Hydrogen(CMakePackage, CudaPackage):
             description='Builds with support for FP16 precision data types')
 
     conflicts('~openmp', when='+omp_taskloops')
+    conflicts('+cuda', when='+rocm')
+    conflicts('+rocm', when='+cuda')
```
```suggestion
    conflicts('+cuda', when='+rocm', msg='CUDA and ROCm support are mutually exclusive')
```

haampie commented on var/spack/repos/builtin/packages/lbann/package.py:
```diff
@@ -71,6 +71,8 @@ class Lbann(CMakePackage, CudaPackage):
     conflicts('~cuda', when='+nvprof')
     conflicts('~hwloc', when='+al')
     conflicts('~cuda', when='+nvshmem')
+    conflicts('+cuda', when='+rocm')
+    conflicts('+rocm', when='+cuda')
```
```suggestion
    conflicts('+cuda', when='+rocm', msg='CUDA and ROCm support are mutually exclusive')
```

bvanessen committed:
Apply suggestions from code review

Co-authored-by: Harmen Stoppels <harmenstoppels@gmail.com>

bvanessen committed:
Added the ability to use the hipcc compiler when the ROCm package is
enabled.  Also removed LBANN CUDA CMake flags that are set by the
version of Hydrogen that is compiled against.

haampie commented:
@bvanessen I haven't really checked out the packages, but if possible (when device / host code is well separated) you might want to drop the CXX=hipcc variable and leave it up to cmake to use hipcc for sources with device code (like it is with cuda: if the language is enabled, nvcc is applied to just .cu files by default). That way the user can still choose to compile host code with e.g. gcc, or they can register hipcc as a compiler and compile everything with it.

bvanessen committed:
Updates to ensure that LBANN software stack builds with ROCm / HIP
software stack.  Updated recipes to use HWLOC 2.3 with ROCm to enable
topology awareness.

bvanessen committed:
Simplified the ROCm / HIP dependencies based on updates to the HIP
package.

bvanessen commented:
Depends on improvements in the HIP package in #21822

bvanessen committed:
Flake8

bvanessen committed:
Explicitly set the HIP CXX compiler so that the HIP package can figure things out

haampie commented:
Thanks @bvanessen, does this mean this PR is good to go, we could close #21822, and try to get #21852 in?

bvanessen commented:
Yes we are good to go.  I will close 21822 and we can go from 21852.

Brian C. Van Essen
vanessen1@llnl.gov
(w) 925-422-9300
(c) 925-290-5470





> On Feb 22, 2021, at 1:42 PM, Harmen Stoppels <notifications@github.com> wrote:
> 
> 
> Thanks @bvanessen, does this mean this PR is good to go, we could close #21822, and try to get #21852 in?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.
>

bvanessen commented:
@haampie Thanks.
